services/src/main/java/org/jboss/windup/web/services/service/PackageService.java

```diff
@@ -0,0 +1,128 @@
+import org.jboss.windup.web.services.model.RegisteredApplication;
+
+/**
+ * Service for manipulation with packages
+ *
+ * @author <a href="mailto:dklingenberg@gmail.com">David Klingenberg</a>
+ */
+public class PackageService
+{
+    @PersistenceContext
+    EntityManager entityManager;
+
+    @Inject
+    private Furnace furnace;
+
+    @Inject
+    @FromFurnace
+    private PackageDiscoveryService packageDiscoveryService;
+
+    /**
+     * Discovers packages in application
+     *
+     * @param application Application to discover packages in
+     * @return Collection of discovered packages
+     */
+    public Collection<Package> discoverPackages(RegisteredApplication application)
+    {
+        PackageMetadata metadata = application.getPackageMetadata();
+        metadata.setScanStatus(PackageMetadata.ScanStatus.IN_PROGRESS);
+        this.entityManager.merge(metadata);
+
+        ApplicationGroup appGroup = application.getApplicationGroup();
+        PackageMetadata appGroupMetadata = appGroup.getPackageMetadata();
+        appGroupMetadata.setScanStatus(PackageMetadata.ScanStatus.IN_PROGRESS);
+        this.entityManager.merge(appGroupMetadata);
+
+        String inputPath = application.getInputPath();
+        PackageDiscoveryService.PackageDiscoveryResult result = this.packageDiscoveryService.execute(inputPath);
+
+        Map<String, Package> packageMap = new TreeMap<>();
+
+        this.addPackagesToPackageMetadata(metadata, result.getKnownPackages(), packageMap);
+        this.addPackagesToPackageMetadata(metadata, result.getUnknownPackages(), packageMap);
+
+        this.addPackagesToPackageMetadata(appGroupMetadata, result.getKnownPackages(), packageMap);
+        this.addPackagesToPackageMetadata(appGroupMetadata, result.getUnknownPackages(), packageMap);
+
+        metadata.setScanStatus(PackageMetadata.ScanStatus.COMPLETE);
+        this.entityManager.merge(metadata);
+
+        PackageMetadata.ScanStatus finalStatus = appGroup.getApplications().stream()
+                    .map(app -> app.getPackageMetadata().getScanStatus())
+                    .reduce(PackageMetadata.ScanStatus.COMPLETE,
+                                (a, b) -> a == b ? PackageMetadata.ScanStatus.COMPLETE : PackageMetadata.ScanStatus.IN_PROGRESS);
+
+        appGroupMetadata.setScanStatus(finalStatus);
+        this.entityManager.merge(appGroupMetadata);
+
+        return metadata.getPackages();
+    }
+
+    private void addPackagesToPackageMetadata(PackageMetadata metadata, Map<String, Integer> discoveredPackages, Map<String, Package> packageMap)
+    {
+        for (String packageName : discoveredPackages.keySet())
+        {
+            Package aPackage = this.createPackage(packageName, discoveredPackages, packageMap);
+            metadata.addPackage(aPackage);
+        }
+    }
+
+    Package createPackage(String fullPackageName, Map<String, Integer> packageClassCount, Map<String, Package> packageMap)
```

Comment:
I convinced myself that this works right if you add "x.y.a" and then "x.y.b" to the same map, but on first read I thought the x.y.b Package would be missing its parent. Might be good to have a test showing that it works correctly with a pair of packages sharing the first part of their name, then the reader won't have to convince himself (and it won't be as easy to accidentally break that case.)

(It may be easier to test this case through addPackagesToPackageMetadata?)

Anyway, I will leave the reviewing to @jsight, just wanted to check out this PR and it's looking good.